src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -69,136 +75,174 @@ export class KernelFinder implements IKernelFinder {
+        if (preferred >= 0) {
+            return kernels[preferred];
         }
     }
+
     // Search all our local file system locations for installed kernel specs and return them
     @captureTelemetry(Telemetry.KernelListingPerf)
-    public async listKernelSpecs(resource: Resource): Promise<IJupyterKernelSpec[]> {
+    public async listKernels(resource: Resource): Promise<LocalKernelConnectionMetadata[]> {
         // Get an id for the workspace folder, if we don't have one, use the fsPath of the resource
         const workspaceFolderId = this.workspaceService.getWorkspaceFolderIdentifier(
             resource,
             resource?.fsPath || this.workspaceService.rootPath
         );
 
         // If we have not already searched for this resource, then generate the search
-        if (workspaceFolderId && !this.workspaceToKernels.has(workspaceFolderId)) {
-            this.workspaceToKernels.set(workspaceFolderId, this.findResourceKernelSpecs(resource));
+        if (workspaceFolderId && !this.workspaceToMetadata.has(workspaceFolderId)) {
+            this.workspaceToMetadata.set(workspaceFolderId, this.findResourceKernelMetadata(resource));
         }
 
         this.writeCache().ignoreErrors();
 
         // ! as the has and set above verify that we have a return here
-        const promise = this.workspaceToKernels.get(workspaceFolderId)!;
-        promise
-            .then((items) =>
-                traceInfoIf(
-                    !!process.env.VSC_JUPYTER_LOG_KERNEL_OUTPUT,
-                    `Kernel specs for ${resource?.toString() || 'undefined'} are \n ${JSON.stringify(items)}`
-                )
-            )
-            .catch(noop);
-        return promise;
+        const promise = this.workspaceToMetadata.get(workspaceFolderId)!;
+        return promise.then((items) => {
+            traceInfoIf(
+                !!process.env.VSC_JUPYTER_LOG_KERNEL_OUTPUT,
+                `Kernel specs for ${resource?.toString() || 'undefined'} are \n ${JSON.stringify(items)}`
+            );
+            return items;
+        });
     }
 
-    private async findKernelSpecBasedOnNotebookMetadata(
-        resource: Resource,
-        notebookMetadata?: nbformat.INotebookMetadata
-    ) {
-        traceInfo(
-            `Searching for kernel based on ${JSON.stringify(notebookMetadata?.kernelspec || {})} for ${
-                resource?.fsPath || ''
-            }`
-        );
-        await this.readCache();
-
-        const searchBasedOnKernelSpecMetadata = this.findKernelSpecBasedOnKernelSpecMetadata(
-            resource,
-            notebookMetadata && notebookMetadata.kernelspec ? notebookMetadata.kernelspec : undefined
-        );
-
-        if (!notebookMetadata || notebookMetadata.kernelspec || !notebookMetadata.language_info?.name) {
-            return searchBasedOnKernelSpecMetadata;
+    public async getKernelSpecRootPath(): Promise<string | undefined> {
```

Comment:
This doesn't need to be async.

Reply:
Not sure about these paths, I think the correct ones are the ones in line 35,36,
These to my knowledge are the correct global root paths for kernel spec (based on their docs & what we have used today).
```typescript
const winJupyterPath = path.join('AppData', 'Roaming', 'jupyter', 'kernels');
const linuxJupyterPath = path.join('.local', 'share', 'jupyter', 'kernels');
```

Reply:
The real path has to be passed back, not the appdata roaming one.

Reply:
Yes, but if you look at line 138, 
we're returning `return path.join('/', 'usr', 'share', 'jupyter', 'kernels');`
when it should be one of the two based on the OS:
```
const linuxJupyterPath = path.join('.local', 'share', 'jupyter', 'kernels');
const macJupyterPath = path.join('Library', 'Jupyter', 'kernels');
```

Reply:
If you look at line 525 in localkernelfinder.ts you'll see that mac and linux always start with /usr/share/jupyter/kernels. That path is the same between them both.

The docs for jupyter mention the same thing:
https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs

I'll add a comment to that effect.

Reply:
Sure, but this crashes and burns, for me (due to permission error, i cannot write to those directories in a mac).
Besides the old code also uses the path i have mentioned as well, in line 525.
I think reading from there is fine, but when writing (creating the file) on a Mac, we must use the user version as it fails for me, & it could fail for others too.

I think we should do the same for Linux as well, as its possible users won't have permissions to write into the `/usr/...` directory

Reply:
Ah yeah that makes more sense. User folder should be the default because we should be able to write to it.